protocols/kad/src/handler.rs

```diff
@@ -970,13 +989,12 @@ where
 
                     return Poll::Ready(Some(ConnectionHandlerEvent::Custom(event)));
                 }
-                OutboundSubstreamState::Closing(mut stream) => match stream.poll_close_unpin(cx) {
-                    Poll::Ready(Ok(())) | Poll::Ready(Err(_)) => return Poll::Ready(None),
-                    Poll::Pending => {
-                        *this = OutboundSubstreamState::Closing(stream);
-                        return Poll::Pending;
-                    }
-                },
+                OutboundSubstreamState::Available(substream) => {
+                    *this = OutboundSubstreamState::Done;
+                    let event = KademliaHandlerEvent::Available { substream };
```

Comment:
Instead of returning the stream here, can we just remain in `Available` and search all streams for one of these before we open a new one (assuming we are still within the limit)?

That should avoid the `unreachable` case in the behaviour I think.

Reply:
I considered that, you mentioned such approach in https://github.com/libp2p/rust-libp2p/discussions/3411. I didn't like it because it is a hot path and it should be slower to iterate streams in `SelectAll` one by one rather than taking one deterministically from a stored vector. I don't have a very strong opinion here though if that is what you'd like to do anyway.